scripts/abins/sdata.py

```diff
@@ -4,78 +4,123 @@
+                 ) -> None:
+        super().__init__()
+
+        if temperature is None:
+            self._temperature = None
+        elif isinstance(temperature, Real):
+            self._temperature = float(temperature)
+        else:
+            raise TypeError("Temperature must be a real number or None")
 
-        if sample_form in ALL_SAMPLE_FORMS:
+        if isinstance(sample_form, str):
             self._sample_form = sample_form
         else:
-            raise ValueError("Invalid sample form %s" % sample_form)
+            raise TypeError("Sample form must be a string. Use '' (default) if unspecified.")
 
-        self._data = None  # dictionary which stores dynamical structure factor for all atoms
-        self._bin_width = None
+        self._frequencies = np.asarray(frequencies, dtype=FLOAT_TYPE)
+        self._check_frequencies()
 
-    def set_bin_width(self, width=None):
-        self._bin_width = width
+        self._data = data
+        self._check_data()
 
-    def set(self, items=None):
-        """
-        Sets a new value for a collection of the data.
+    def get_frequencies(self) -> np.ndarray:
+        return self._frequencies.copy()
+
+    def get_temperature(self) -> Union[float, None]:
+        return self._temperature
+
+    def get_sample_form(self) -> str:
+        return self._sample_form
+
+    def get_bin_width(self) -> Union[float, None]:
+        """Check frequency series and return the bin size
+
+        If the frequency series does not have a consistent step size, return None
         """
-        if not isinstance(items, dict):
-            raise ValueError("New value of S  should have a form of a dict.")
 
-        for item in items:
-            if ATOM_LABEL in item:
+        self._check_frequencies()
+        step_size = (self._frequencies[-1] - self._frequencies[0]) / (self._frequencies.size - 1)
+
+        if np.allclose(step_size, self._frequencies[1:] - self._frequencies[:-1]):
+            return step_size
+        else:
+            return None
 
-                if not isinstance(items[item], dict):
+    def check_finite_temperature(self):
+        """Raise an error if Temperature is not greater than zero"""
+        temperature = self.get_temperature()
+        if not (isinstance(temperature, (float, int)) and temperature > 0):
+            raise ValueError("Invalid value of temperature.")
+
+    def check_known_sample_form(self):
+        """Raise an error if sample form is not known to Abins"""
+        sample_form = self.get_sample_form()
+        if sample_form not in ALL_SAMPLE_FORMS:
+            raise ValueError(
+                f"Invalid sample form {sample_form}: known sample forms are {ALL_SAMPLE_FORMS}")
+
+    def _check_frequencies(self):
+        # Check frequencies are ordered low to high
+        if not np.allclose(np.sort(self._frequencies),
+                           self._frequencies):
+            raise ValueError("Frequencies not sorted low to high")
```

Comment:
If you are sorting the array to check if it is sorted anyway, would it be beneficial to correct the array instead of throwing?

Reply:
The scattering data corresponds to the frequency array and would also need to be re-arranged - which is certainly doable. But at this point the error is more of a sanity-check for development. If a Mantid/Abins user sees it, something else has gone badly wrong and I'd like to know about it.

It could make sense to introduce such conveniences if/when we start importing SData from other sources; possibly as an alternate constructor method?

Reply:
I'd prefer to leave this as-is for now. Would revisit if implementing import of data from possibly-unsorted sources.